x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportPutDataFrameJobAction.java

```diff
@@ -68,37 +78,72 @@ protected void masterOperation(Request request, ClusterState clusterState, Actio
+        String jobId = request.getConfig().getId();
+        // quick check whether a job has already been created under that name
+        if (PersistentTasksCustomMetaData.getTaskWithId(clusterState, jobId) != null) {
+            listener.onFailure(new ResourceAlreadyExistsException(
+                    DataFrameMessages.getMessage(DataFrameMessages.REST_PUT_DATA_FRAME_JOB_EXISTS, jobId)));
+            return;
+        }
 
+        // create the job, note the non-state creating steps are done first, so we minimize the chance to end up with orphaned state
+        // job validation
+        JobValidator jobCreator = new JobValidator(request.getConfig(), client);
         jobCreator.validate(ActionListener.wrap(validationResult -> {
+            // deduce target mappings
             jobCreator.deduceMappings(ActionListener.wrap(mappings -> {
-                DataFrameJob job = createDataFrameJob(request.getConfig(), threadPool);
-                DataframeIndex.createDestinationIndex(client, job, mappings, ActionListener.wrap(createIndexResult -> {
-                    startPersistentTask(job, listener, persistentTasksService);
-                }, e3 -> {
-                    listener.onFailure(new RuntimeException("Failed to create index", e3));
+                // create the destination index
+                DataframeIndex.createDestinationIndex(client, request.getConfig(), mappings, ActionListener.wrap(createIndexResult -> {
+                    DataFrameJob job = createDataFrameJob(jobId, threadPool);
+                    // create the job configuration and store it in the internal index
+                    dataFrameJobConfigManager.putJobConfiguration(request.getConfig(), false, ActionListener.wrap(r -> {
+                        // finally start the persistent task
+                        persistentTasksService.sendStartRequest(job.getId(), DataFrameJob.NAME, job, ActionListener.wrap(persistentTask -> {
+                            listener.onResponse(new PutDataFrameJobAction.Response(true));
+                        }, startPersistentTaskException -> {
+                            // delete the otherwise orphaned job configuration, for now we do not delete the destination index
+                            dataFrameJobConfigManager.deleteJobConfiguration(jobId, ActionListener.wrap(r2 -> {
+                                logger.debug("Deleted data frame job [{}] configuration from data frame configuration index", jobId);
+                                listener.onFailure(
+                                        new RuntimeException(
+                                                DataFrameMessages.getMessage(
+                                                        DataFrameMessages.REST_PUT_DATA_FRAME_FAILED_TO_START_PERSISTENT_TASK, r2),
+                                                startPersistentTaskException));
+                            }, deleteJobFromIndexException -> {
```

Comment:
We don't log or return these exception anywhere. 

Do we at least want to log it?

Reply:
we return the exception as part of the rest call, so you get the exception in the Rest response. 

The delete one is a tricky one: If the start of the persistent task fails you have an orphaned configuration/document in the config index. The other option would be to start the task first and store the config second, but if the put config fails, you have the other problem of a task with no job configuration, so you would need to stop the task if storing the config fails. Either way trouble, I decided for storing the config first because I expect that to fail more likely than starting the task. As you can see I am trying to delete the orphaned object, return the exception from the persistent task failure.

I am not deleting the destination index because that isn't really an internal implementation detail and users usually know how to delete an index.

FWIW. I am working on the delete api to delete the configuration document even if the task doesn't exist anymore, so that in case you corrupt your cluster you can get back to a sane state using the data frame API's and not manually fiddling with the configuration index.